src/models/geotimeline.py

```diff
@@ -1,6 +1,522 @@
+
+        Raises:
+            ValueError: If max_travel_distance < 0.
+        """
+
+        for graph in self:
+            graph.add_habitat(
+                name=name,
+                valid_classes=valid_classes,
+                barrier_classes=barrier_classes,
+                max_travel_distance=max_travel_distance,
+                add_distance=add_distance,
+                add_component_edges=add_component_edges,
+            )
+
+        self.habitats[name] = GeoGraphTimeline(
+            {year: graph.habitats[name] for year, graph in self.graphs.items()}
+        )
+
+    def calculate_node_dynamics(self, time: TimeStamp) -> pd.Series:
+        """
+        Classify the dynamic behavior of each node of the graph at the given `time`.
+
+        The node dynamics reflects the type of change that a node has undergone between
+        two adjacent time-slices of a GeoGraphTimeline.
+        For each node at the selected `time` the classified node dynamics is one of:
+            `birth`: The child node has no ancestors in `nodemap` and newly appeared.
+            `split`: The child node was created as a split off of an ancestral node.
+            `unchanged`: The child node has one ancestor with the same characteristics.
+            `grew`: The child node has one ancestor and has increased in area.
+            `shrank`: The child node has one ancestor and has decreased in area.
+            `complex`: The child node has numerous ancestors and siblings
+            `merged`: The child node has several ancestors and no siblings. It was
+                created from a merge of several ancestral nodes.
+
+        Args:
+            time (TimeStamp): The time in GeoGraphTimeline for which the node dynamics
+                should be classified. Must be one of the times in `self.times`.
+
+        Raises:
+            UserWarning: If an inexistent timeslice is accessed.
+
+        Returns:
+            pd.Series: A pandas series with the node dynamics for each node
+                the graph at the given time (self[time]).
+        """
+
+        if "node_dynamic" in self[time].df.columns:
+            return self[time].df["node_dynamic"]
+        else:
+            # Determine index of the selected time in self.times
+            time_index = bisect_left(self.times, time)
+
+            if time_index == 0:
+                raise UserWarning(
+                    "Cannot calculate node dyamics for first graph in timeline."
+                )
+
+            prior_time = self.times[time_index - 1]
+            node_map = self.node_map_cache(prior_time, time)
+
+            # Helper function to calculate dynamics type via a single map call
+            dynamics_oracle = (
+                lambda child_index: GeoGraphTimeline._calculate_node_dynamics(
+                    child_index, node_map
+                )
+            )
+
+            dynamics_type = self[time].df.index.map(dynamics_oracle)
```

Comment:
Does this return a `Series` (so it has loc indexes) or just a List (so it has iloc indexes)?

Reply:
This returns a Series 😉